app/bundles/LeadBundle/Tests/Entity/TagRepositoryTest.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+/*
+ * @copyright   2016 Mautic Contributors. All rights reserved
+ * @author      Mautic, Inc.
+ *
+ * @link        https://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\LeadBundle\Tests\Entity;
+
+use Doctrine\ORM\EntityManager;
+use Mautic\LeadBundle\Entity\Tag;
+use Mautic\LeadBundle\Entity\TagRepository;
+
+class TagRepositoryTest extends \PHPUnit_Framework_TestCase
+{
+    public function testSaveEntityWithUniqueNewTag()
+    {
+        $entity = new Tag('sometag');
+
+        $mockEntityManager = $this->getMockBuilder(EntityManager::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['persist', 'flush'])
+            ->getMock();
+
+        $mockEntityManager->expects($this->once())
+            ->method('persist')
+            ->with($entity);
+
+        $mockRepository = $this->getMockBuilder(TagRepository::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getEntityManager', 'findOneBy'])
+            ->getMock();
+
+        $mockRepository->method('getEntityManager')
+            ->willReturn($mockEntityManager);
+
+        $mockRepository->expects($this->once())
+            ->method('findOneBy')
+            ->with(['tag' => 'sometag'])
+            ->willReturn(null);
+
+        $mockRepository->saveEntity($entity);
```

Comment:
I do not like testing on mocks. I use class directly:
`$tagRepository = new TagRepository(..)`

Reply:
Can you advise how to test that method `findOneBy` is called once with the expected array without mocking the repository?

Reply:
I usually do not use this prepared methods (typehint problem for example), so I call it directly on QB and I can mock it: https://github.com/mautic/mautic/pull/5205/files#diff-3be505232475711cc0511787815d6571

It will be a little bit more work if you use `findOneBy` because you need to mock UnionWork and Persister.

I think we can leave this if you are sure the test will cover functionality, it is about personal preferences maybe.